homeassistant/components/homeassistant/exposed_entities.py

```diff
@@ -0,0 +1,302 @@
+STORAGE_VERSION = 1
+
+SAVE_DELAY = 10
+
+DEFAULT_EXPOSED_DOMAINS = [
+    "climate",
+    "cover",
+    "fan",
+    "humidifier",
+    "light",
+    "lock",
+    "scene",
+    "script",
+    "switch",
+    "vacuum",
+    "water_heater",
+]
+
+DEFAULT_EXPOSED_BINARY_SENSOR_DEVICE_CLASSES = [
+    BinarySensorDeviceClass.DOOR,
+    BinarySensorDeviceClass.GARAGE_DOOR,
+    BinarySensorDeviceClass.LOCK,
+    BinarySensorDeviceClass.MOTION,
+    BinarySensorDeviceClass.OPENING,
+    BinarySensorDeviceClass.PRESENCE,
+    BinarySensorDeviceClass.WINDOW,
+]
+
+DEFAULT_EXPOSED_SENSOR_DEVICE_CLASSES = [
+    SensorDeviceClass.AQI,
+    SensorDeviceClass.CO,
+    SensorDeviceClass.CO2,
+    SensorDeviceClass.HUMIDITY,
+    SensorDeviceClass.PM10,
+    SensorDeviceClass.PM25,
+    SensorDeviceClass.TEMPERATURE,
+    SensorDeviceClass.VOLATILE_ORGANIC_COMPOUNDS,
+]
+
+
+@dataclasses.dataclass(frozen=True)
+class AssistantPreferences:
+    """Preferences for an assistant."""
+
+    expose_new: bool
+
+    def to_json(self) -> dict[str, Any]:
+        """Return a JSON serializable representation for storage."""
+        return {"expose_new": self.expose_new}
+
+
+class ExposedEntities:
+    """Control assistant settings."""
+
+    _assistants: dict[str, AssistantPreferences]
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize."""
+        self._hass = hass
+        self._listeners: dict[str, list[Callable[[], None]]] = {}
+        self._store: Store[dict[str, dict[str, dict[str, Any]]]] = Store(
+            hass, STORAGE_VERSION, STORAGE_KEY
+        )
+
+    async def async_initialize(self) -> None:
+        """Finish initializing."""
+        websocket_api.async_register_command(self._hass, ws_expose_entity)
+        websocket_api.async_register_command(self._hass, ws_expose_new_entities)
+        await self.async_load()
+
+    @callback
+    def async_listen_entity_updates(
+        self, assistant: str, listener: Callable[[], None]
+    ) -> None:
+        """Listen for updates to entity expose settings."""
+        self._listeners.setdefault(assistant, []).append(listener)
+
+    @callback
+    def async_expose_entity(
+        self, assistant: str, entity_id: str, should_expose: bool
+    ) -> None:
+        """Expose an entity to an assistant.
+
+        Notify listeners if expose flag was changed.
+        """
+        entity_registry = er.async_get(self._hass)
+        if not (registry_entry := entity_registry.async_get(entity_id)):
+            return
+
+        if (
+            assistant_options := registry_entry.options.get(assistant)
+        ) and assistant_options["should_expose"] == should_expose:
+            return
+
+        assistant_options = {"should_expose": should_expose}
+        entity_registry.async_update_entity_options(
+            entity_id, assistant, assistant_options
+        )
+        for listener in self._listeners.get(assistant, []):
+            listener()
```

Comment:
Should we pass in the changed info ? Alexa for example can update individual entities.